Source/WebKit/webpushd/mac/com.apple.WebKit.webpushd.sb.in

```diff
@@ -22,8 +22,308 @@
+
+;; Read-only preferences and data
+(allow file-read*
+    ;; Basic system paths
+    (subpath "/Library/Frameworks")
+    (subpath "/Library/Managed Preferences")
+    (regex #"^/private/etc/(hosts|group|passwd)$")
+
+    ;; System and user preferences
+    (home-literal "/.CFUserTextEncoding"))
+
+(allow file-read-data
+    (literal "/usr/local/lib/log")) ; <rdar://problem/36629495>
+
+;; Preferences support
+(shared-preferences-read
+    "com.apple.LaunchServices"
+    "com.apple.crypto")
+
+(allow-reading-global-preferences)
+
+; FIXME: This is needed for some security framework calls (that use non-CFPreferences readers)
+(allow file-read-data
+    (literal "/Library/Preferences/com.apple.security.plist")
+    (home-subpath "/Library/Preferences/com.apple.security.plist"))
+
+;; On-disk WebKit2 framework location, to account for debug installations outside of /System/Library/Frameworks,
+;; and to allow issuing extensions.
+(allow-read-directory-and-issue-read-extensions (param "WEBKIT2_FRAMEWORK_DIR"))
+
+#if USE(SYSTEM_CONTENT_PATH)
+(map allow-read-directory-and-issue-read-extensions
+    read-directory-and-issue-read-extension-secondary-paths)
+#endif
+
+(allow mach-lookup
+#if PLATFORM(MAC)
+    (with report) (with telemetry)
+#endif
+#if !ENABLE(CFPREFS_DIRECT_MODE)
+    (global-name "com.apple.cfprefsd.agent")
+#endif
+    (global-name "com.apple.system.logger")
+)
+
+;; Various services required by AppKit and other frameworks
+(allow mach-lookup
+    (global-name "com.apple.SystemConfiguration.configd")
+#if !ENABLE(CFPREFS_DIRECT_MODE)
+    (global-name "com.apple.cfprefsd.daemon")
+#endif
+    (global-name "com.apple.logd")
+    (global-name "com.apple.logd.events")
+)
+
+(deny mach-lookup (with no-log)
+    (global-name "com.apple.CoreServices.coreservicesd")
+    (global-name "com.apple.DiskArbitration.diskarbitrationd")
+    (global-name "com.apple.ViewBridgeAuxiliary")
+    (global-name "com.apple.system.opendirectoryd.membership"))
+
+;; CoreFoundation. We don't import com.apple.corefoundation.sb, because it allows unnecessary access to pasteboard.
+(allow mach-lookup
+    (global-name-prefix "com.apple.distributed_notifications")
+#if !HAVE(CSCHECKFIXDISABLE)
+    (global-name "com.apple.CoreServices.coreservicesd")
+#endif
+)
+
+(deny file-write-create (vnode-type SYMLINK))
+
+;; Reserve a namespace for additional protected extended attributes.
+(deny file-read-xattr file-write-xattr (xattr-prefix "com.apple.security.private."))
+
+;; Analytics.
+(allow mach-lookup (global-name "com.apple.analyticsd"))
+(allow mach-lookup (global-name "com.apple.awdd"))
+
+;; Daemon prefs.
+(allow user-preference-read user-preference-write
+    (preference-domain "com.apple.webkit.webpushd"))
+
+;; Push notification registration.
+(allow mach-lookup
+    (global-name "com.apple.apsd")
+    (global-name "com.apple.windowserver.active")) ;; for -[APSSystemUser isLoggedIn]
```

Comment:
Is it possible to do without access to the WindowServer? In other profiles, we block this and call CGSSetDenyWindowServerConnections.

Reply:
I don't think this would work with the way `APSSystemUser` is implemented. They call `CGSCopySessionList` to determine whether the current user id is logged in and as far as I can tell that requires messaging WindowServer. If you block the WindowServer connection then the APS client code will think that the current user is logged out.

We could file a bug on that component to see if they can implement this logic some other way and modify our sandbox profile to block the WS connection after that's been fixed.

Reply:
Yes, that makes sense, thanks!